src/coreComponents/physicsSolvers/wavePropagation/ElasticWaveEquationSEMKernel.hpp

```diff
@@ -0,0 +1,723 @@
+
+      //Test to make sure if the normal is outwardly directed
+      LvArray::tensorOps::copy< 3 >( tmpVector, faceCenterOnFace );
+      LvArray::tensorOps::subtract< 3 >( tmpVector, elemCenter );
+      if( LvArray::tensorOps::AiBi< 3 >( tmpVector, faceNormalOnFace ) < 0.0 )
+      {
+        LvArray::tensorOps::scale< 3 >( faceNormalOnFace, -1 );
+      }
+
+      // compute the vector face center to query point
+      LvArray::tensorOps::subtract< 3 >( faceCenterOnFace, coords );
+      localIndex const s = computationalGeometry::sign( LvArray::tensorOps::AiBi< 3 >( faceNormalOnFace, faceCenterOnFace ));
+
+      // all dot products should be non-negative (we enforce outward normals)
+      if( prevSign * s < 0 )
+      {
+        return false;
+      }
+      prevSign = s;
+
+    }
+    return true;
+  }
+
+  /**
+   * @brief Convert a mesh element point coordinate into a coordinate on the reference element
+   * @tparam FE_TYPE finite element type
+   * @param[in] numFacesPerElem number of face on an element
+   * @param[in] coords coordinate of the point
+   * @param[in] elemCenter array containing the center of the elements
+   * @param[in] faceNormal array containing the normal of all faces
+   * @param[in] faceCenter array containing the center of all faces
+   * @param[in] elemsToNodes map to obtaint global nodes from element index
+   * @param[in] elemsToFaces map to get the global faces from element index and local face index
+   * @param[in] X array of mesh nodes coordinates
+   * @param[out] coordsOnRefElem to contain the coordinate computed in the reference element
+   * @return true if coords is inside the element num index
+   */
+  template< typename FE_TYPE >
+  GEOSX_HOST_DEVICE
+  static void
+  computeCoordinatesOnReferenceElement( real64 const (&coords)[3],
+                                        arraySlice1d< localIndex const, cells::NODE_MAP_USD - 1 > const elemsToNodes,
+                                        arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const X,
+                                        real64 (& coordsOnRefElem)[3] )
+  {
+    real64 xLocal[FE_TYPE::numNodes][3]{};
+    for( localIndex a = 0; a < FE_TYPE::numNodes; ++a )
+    {
+      LvArray::tensorOps::copy< 3 >( xLocal[a], X[ elemsToNodes[a] ] );
+    }
+    // coordsOnRefElem = invJ*(coords-coordsNode_0)
+    real64 invJ[3][3]{};
+    FE_TYPE::invJacobianTransformation( 0, xLocal, invJ );
+    for( localIndex i = 0; i < 3; ++i )
+    {
+      // init at (-1,-1,-1) as the origin of the referential elem
+      coordsOnRefElem[i] = -1.0;
+      for( localIndex j = 0; j < 3; ++j )
+      {
+        coordsOnRefElem[i] += invJ[i][j] * (coords[j] - xLocal[0][j]);
+      }
+    }
+  }
+
+  GEOSX_HOST_DEVICE
+  static real64
+  evaluateRicker( real64 const & time_n,
+                  real64 const & f0,
+                  localIndex const & order )
+  {
+    real64 const o_tpeak = 1.0/f0;
```

Comment:
Are we sure that this can never ever be zero ?

Reply:
It can in the case someone does not know how to use the source stuff
BY the way all those kind of source stuff will be refactor later maybe a the same time we can put a test to check f0 value and threw out the user if it is set to 0 inside the XMl (the setting itself is mandatory by the way).